share/models/meta.py

```diff
@@ -50,28 +52,67 @@ class Disambiguation:
         all = ('name',)
 
 
-class SubjectManager(FuzzyCountManager):
-    def get_by_natural_key(self, subject):
-        return self.get(name=subject)
+class SubjectTaxonomy(models.Model):
+    name = models.TextField(unique=True)
+    is_deleted = models.BooleanField(default=False)
+    date_created = models.DateTimeField(auto_now_add=True)
+    date_modified = models.DateTimeField(auto_now=True)
 
+    def __str__(self):
+        return self.name
 
-class Subject(models.Model):
-    parent = models.ForeignKey('self', null=True, on_delete=models.CASCADE)
-    name = models.TextField(unique=True, db_index=True)
+    def __repr__(self):
+        return '<{}: {}>'.format(self.__class__.__name__, self.name)
 
-    objects = SubjectManager()
+    class Meta:
+        verbose_name_plural = 'Subject Taxonomies'
 
-    def __str__(self):
-        return self.name
 
-    def natural_key(self):
+class Subject(ShareObject):
+    name = models.TextField()
+    is_deleted = models.BooleanField(default=False)
+    uri = ShareURLField(unique=True, null=True, blank=True)
+    taxonomy = models.ForeignKey(SubjectTaxonomy, editable=False, on_delete=models.CASCADE)
+    parent = ShareForeignKey('Subject', blank=True, null=True, related_name='children')
+    central_synonym = ShareForeignKey('Subject', blank=True, null=True, related_name='custom_synonyms')
+
+    @classmethod
+    def normalize(cls, node, graph):
+        edge = node.related('central_synonym')
+        if edge and edge.related and edge.related.id == node.id:
+            graph.remove_edge(edge)
+
+    def save(self, *args, **kwargs):
+        if self.id is not None and self.parent is not None:
+            new_lineage = self.parent.lineage()
+            if self in new_lineage:
```

Comment:
Preventing cycles seemed important enough to guarantee a check on every save... where do you think would make more sense?